docs/architecture/adr-006-subspace-module.md

```diff
@@ -0,0 +1,221 @@
+  
+  // Represents the account that is associated with the subspace and
+  // should be used to connect external applications to verify this subspace
+  Treasury string
+
+  // the creation time of the subspace
+  CreationTime time.Time
+}
+```
+
+#### ACL
+In order to easily implement an ACL, we will use a simple set of keys made as follows: 
+```
+ACLPrefix + Subspace ID + <Group name | User Address> -> ACL Value
+```
+
+The `ACL Value` will be a simple binary value allowing us to perform bitwise operations to combine the following different permissions: 
+
+```go
+const (
+  // Identifies users that can create content inside the subspace
+  PermissionWrite           = 0b000001
+  
+  // Allows users to moderate contents of other users (e.g. deleting it) 
+  PermissionModerateContent = 0b000010
+  
+  // Allows to add a link for this subspace
+  PermissionAddLink         = 0b000100
+  
+  // Allows to change the information of the subspace
+  PermissionChangeInfo      = 0b001000
+  
+  // Allows to set other users' permissions (except PermissionSetPermissions). 
+  // This includes managing user groups and the associated permissions
+  PermissionSetPermissions  = 0b010000
+)
+```
+
+> **Note**:  
+> Only the `Owner` account will be able to grant other users the `PermissionSetPermissions`
+
+Using this kind of permissions will allow us to easily set permissions and check whether a user has a permission or not: 
+```go
+userPermissions := PermissionWrite | PermissionAddLink | PermissionChangeInfo
+
+canWrite := (userPermissions & PermissionWrite) == PermissionWrite  // True
+canModerateContent := (userPermissions & PermissionModerateContent) == PermissionModerateContent // False
+```
+
+##### Group permissions 
+In order to simplify the handling of multiple users' permissions, we SHOULD allow subspace admins to set group-wise permissions. 
+Each group will be represented by its own name, which can be defined by the admins themselves. We reserve the group name `Others` to identify all the users that are not part of any other group (i.e. those users who are not registered inside a subspace).
+
+Group permissions MUST be checked after checking the existence of any user permission, in the case that the user has not a more strict permission set.
+
+In order to store the belonging of a user to a group, we will use the following store key and value: 
+```
+GroupPrefix + Subspace ID + User Address + Group name -> 0x01
+```
+The `0x01` value is used here only to signal that the specific user is part of that group.
+This will allow us to iterate over all the groups that a user is part of based on their address.
+
+### `Msg` Service
+We will allow the following operations to be performed.
+
+**Subspace administration**
+* Create a subspace
+* Edit a subspace
+
+**Content management**
+* Delete contents that do not respect the ToS
+
+**User management**
+* Create a new group
+* Delete a group
+* Set group permissions
+* Set user permissions
```

Comment:
Is that `Set user permissions` including `Add user to a group`? If not, I think we should add it.

Reply:
Thanks for pointing this out. I should have added all the methods necessary. Let me know if something else is missing